### PR TITLE
Apply system's fractional font scaling parameter when displaying text.

### PR DESCRIPTION
**Pull request by iwbnwif**
Use the system font scaling parameter (e.g. "Fonts->Scaling Factor" in Gnome Tweaks, "Force font DPI" in KDE System Settings, GDK_DPI_SCALE environment variable etc.) to scale text before drawing.

This ensures that text in generic controls is drawn at the same size as native controls.

Edit: Please see the mailing list discussion for the background to this PR: https://groups.google.com/d/topic/wx-dev/oYE-bkdPq6g/discussion

### CODE REVIEW

**Comment by iwbnwif** [comment]
Sorry, I forgot that git commits are multi-line and the first line should be a short title :(

**Comment by vadz** [comment]
Hmm, so this broke a bunch of unit tests and I'm not sure why. This definitely needs to be looked into/debugged, but I won't be able to do it today.

**Comment by iwbnwif** [comment]
> Hmm, so this broke a bunch of unit tests and I'm not sure why. This definitely needs to be looked into/debugged, but I won't be able to do it today.

Yes, I saw that and it also breaks wxSTC - the demo in samples/stc/stctest scales the font to double the size.

I suspect that some other things are calling GetNativeFontInfo and expecting an unscaled version of the font.

Thinking on this further, I suggest to grab the scale in one (or more) appropriate wxCairoContext::CreateContext and hold it in a member variable for the lifetime of that context. We can then scale directly at the point of rendering which means the font size remains exactly as now everywhere else.

I note your comment on the mailing list:

> Do you have any experience with changing the resolution while the program
> is running? As I wrote in my previous message in this thread, I think we
> ought to monitor the resolution property for changes, but I haven't
> actually tested this.

My proposal won't implicitly allow for on-the-fly changes of scale. However, I don't think this is a problem because (from memory) both Gnome and KDE warn that font scaling changes will only apply to new instances / restarts of an application.

I will look at this again later today.

**Comment by iwbnwif** [comment]
Agh!! Another battle with Git!! I really don't know what happened here - fast forward kept failing. In the end I had to pull without fast forward, which seems to have created a further commit from June as well as reverting the changes that I had intended to commit.

**Comment by vadz** [comment]
Don't worry about the commit history, I'll just squash-merge this at the end anyhow. If you want to avoid the problems the next time, just don't merge anything into your working branch, only rebase it (if necessary).

There is still one test failure, which could be actually a problem in the test itself, but it would still need to be fixed. Could you please look at it? TIA!

**Comment by iwbnwif** [comment]
There is still one test failure, which could be actually a problem in the test itself, but it would still need to be fixed. Could you please look at it? TIA!

It's a problem with wxControl::Ellipsize with a torture string "a\t\n&string\t\n&with\t\n&many\t\n&chars" and wxELLIPSIZE_FLAGS_EXPAND_TABS set.

It seems that under certain conditions wxControl::Ellipsize decides not to ellipsize the string when it should. 

The tests all pass on my system on normal use, but I I can force it to fail by setting GDK_DPI_SCALE to 2.0.

I suspect a bug in wxControl::Ellipsize, but will investigate more.

**Comment by iwbnwif** [comment]
The AppVeyor failure does not appear to relate to this change at all:
REQUIRE( stream_in.IsOk() )

**Comment by vadz** [comment]
Yes, please ignore the error in `socketStream` test, it happens on AppVeyor from time to time and I couldn't debug why so far.

I don't see anything wrong in neither the test case itself nor in `wxELLIPSIZE_FLAGS_EXPAND_TABS` handling in `wxControlBase::Ellipsize()` though, i.e. neither seems to be hardcoding any pixel values or anything like this, so I still don't understand why does this one fail.

**Comment by iwbnwif** [comment]
Thanks, I will carry on looking later today. One thing is puzzling me - I traced ::GetCharWidth to be using "g" (lower case G), where as I would have thought "W" (upper case W) would be better (see https://stackoverflow.com/questions/3949422/which-letter-of-the-english-alphabet-takes-up-most-pixels).

So, if the string is wider than "g" on average (the test just multiplies GetCharWidth by a fixed amount), then it will fail.

Of course, everything is being scaled equally, so this shouldn't be an issue. But curious nonetheless.

**Comment by iwbnwif** [comment]
This comment is partly intended to be a note to come back to.

The problem may be in src/common/ctrlcmn.cpp EllipsizeCalculator. 

Test string is "&string\t" - this is calculated as 64px at 125% scaling and "Sans 10" font. 64px being the next available tab stop because the text "&string" overflows the first 32px tab stop.

If wxELLIPSIZE_FLAGS_PROCESS_MNEMONICS is set then the code will reduce the overall width by the width of the leading ampersand (e.g. - 12px = 52px). EllipsizeCalculator then returns  true from 
EllipsizationNotNeeded because 52px < 60px (in my example).

But this is not correct, because the remaining "string" also overflows the 32px tab stop, and so the width should still be 64px.

The validation calculation in the test case uses wxControl::RemoveMnemonics to firstly remove the ampersand (resulting in "string\t", then calls dc.GetMultiLineTextExtent which correctly calculates the width to be 64px because as mentioned "string" overflows the first 32px tab stop.

I don't understand why a copy cannot be made of the string in the EllipsizeCalculator ctor, any single ampersands removed and the correct widths calculated. Also, though I don't understand why the test is failing if the font scale is 1.0 because the tab underflow / overflow should be the same (however, the test passes on my system with a font scale of 1.0!).

Adding a getenv to the test case to print the scale.

**Comment by iwbnwif** [comment]
Quick follow up to the previous post... tabs do not appear to be scaled in the same proportion as text.
'W' has width 12px at scale 1.0 and 25px at scale 2.0. '\t' has scale 24px at scale 1.0 and 56px at scale 2.0. This explains why it fails at some scales and passes at others.

Nonetheless, it still looks like EllipsizeCalculator causes the test failure by subtracting the ampersand width *after* the tab has been calculated.

**Comment by vadz** [comment]
> tabs do not appear to be scaled in the same proportion as text.

Thanks, I had no idea. I also didn't know scaling was used at all in the CI builds, in fact.

But this still doesn't really explain why does `Ellipsize()` fail, i.e. why does it produce a string longer than the max width specified when calling it. Do we compute tab widths differently in different places somehow?

**Comment by iwbnwif** [comment]
> But this still doesn't really explain why does Ellipsize() fail, i.e. why does it produce a string longer than the max width specified when calling it. Do we compute tab widths differently in different places somehow?

I think the mechanism is along these lines:
```
|-----t-----t-----t----
 &string\t  | <-width is to here e.g. 2nd tab stop (e.g. 64px).
```

EllipsizeCalculator() simply deducts the space taken up by the ampersand (12px) to arrive at a width of 52px for the above string. 52px is < the max specified 54px (in the AppVeyor context), so is not ellipsized.

The test case then uses wxControl::RemoveMnemonic() so the string becomes like this:

```
|-----t-----t-----t----
 string\t   | <- width is still to 2nd tab stop so **still 64px** (as measured by dc.GetMultiLineTextExtent
```
64px > the max specified 54px so the test fails with 64px > 54px.

In a sense, the test fails because it is not comparing like with like. 

Now, why it passes without my changes is still unresolved! I can either try and fix EllipsizeCalculator() [risky!] or do not make any font scaling calls if the scale is 1.

**Comment by vadz** [comment]
OK, I see, thanks, so this has indeed nothing to do with your changes and is just a bug in the not using `wxELLIPSIZE_FLAGS_EXPAND_TABS` case.

The comment in `EllipsizeCalculator` says that we can't do things differently unless we completely change the code structure, but I'm afraid that we indeed need to do just this. I'll try to do it myself but if you would like to try to fix this, it would be, of course, most welcome too.

The trouble is that I can't merge this PR as we really want the unit test suite to pass, even if, again, the bug is totally unrelated to it and it just happens to expose it.

**Comment by vadz** [comment]
OTOH I don't understand why was c2e8153 ("Revert changes to wxNativeFontInfo::SetFractionalPointSize and implement directly in wxCairoContext instead") made?

**Comment by iwbnwif** [comment]
> OTOH I don't understand why was c2e8153 ("Revert changes to wxNativeFontInfo::SetFractionalPointSize and implement directly in wxCairoContext instead") made?

Sorry, I really don't  know what happened with that Git commit. The majority of those changes are nothing to do with me. 

My commit was to revert the change where I added the new API to src/unix/fontutils.cpp (which caused the original bunch of build failures) and instead to implement them in the wxCairoContext instead - e.g. immediately prior to drawing / measuring the text.

**Comment by vadz** [comment]
I think it could be best to just close this PR and open a new one because things are starting to get really confusing.

But the real question/problem is that I've totally missed that you changed the PR with this commit. I.e. I didn't realize that the initial (simple) approach resulted in many tests (you wrote "build", but you meant "tests", right?) failures. Do you have any idea about why did this happen? Is this unavoidable, i.e. is the initial approach with just modifying `SetFractionalPointSize()` really unworkable? Is this due to the font being scaled twice when used with the native controls in this case perhaps?

BTW, were you able to find out, with the extra debug logs you added, what scaling is used during the CI builds? It would at least help for reproducing the various problems locally. TIA!

**Comment by iwbnwif** [comment]
> OK, I see, thanks, so this has indeed nothing to do with your changes and is just a bug in the not using wxELLIPSIZE_FLAGS_EXPAND_TABS case.

Correct (well that and _using_ wxELLIPSIZE_FLAGS_PROCESS_MNEMONICS). But I would still like to understand why it is passing without my changes. For some reason (probably rounding?) "&string\t" is either being ellipsized or is measuring as < 54px in Travis without my changes.

The problem is I really need to have consistent behavior between Travis and my system to debug this.

> The comment in EllipsizeCalculator says that we can't do things differently unless we completely change the code structure, but I'm afraid that we indeed need to do just this. I'll try to do it myself but if you would like to try to fix this, it would be, of course, most welcome too.

I don't see why we can't make a copy of the string, remove all the ampersands (if wxELLIPSIZE_FLAGS_PROCESS_MNEMONICS is set) and measure the result. Is it concerns about slowing performance with the string copy?

**Comment by iwbnwif** [comment]
> I think it could be best to just close this PR and open a new one because things are starting to get really confusing.

:)) good idea - will do!

> But the real question/problem is that I've totally missed that you changed the PR with this commit. I.e. I didn't realize that the initial (simple) approach resulted in many tests (you wrote "build", but you meant "tests", right?) failures. 

Sorry, yes unit tests. See the above comment [here](https://github.com/wxWidgets/wxWidgets/pull/1635#issuecomment-549263826). 

> Do you have any idea about why did this happen? Is this unavoidable, i.e. is the initial approach with just modifying `SetFractionalPointSize()` really unworkable? Is this due to the font being scaled twice when used with the native controls in this case perhaps?

I think it is unworkable because (IIUC) it is a public function and it seems other aspects of the library (and possibly user applications) are using it without expecting the font size to be scaled - at least at the point they are using it. I really think it is better to keep the font size static within user apps if possible and perform the scaling immediately prior to writing pixels to the screen (or as close as we can control).

> BTW, were you able to find out, with the extra debug logs you added, what scaling is used during the CI builds? It would at least help for reproducing the various problems locally. TIA!

Inclusive, I am afraid - but GDK_DPI_SCALE is unset. I don't know why some Travis tests have 6 x "g" measured as 54px and sometimes as 60px (i.e. 7949.4 vs 7949.5) or is it GTK2 vs GTK3... 

TBH we should really upgrade the tests to inject different font scaling values - but I am too scared to even try that at this stage!!

File: src/generic/graphicc.cpp [hunk 1]
```diff
@@ -2811,7 +2811,7 @@ void wxCairoContext::GetTextExtent( const wxString &str, wxDouble *width, wxDoub
             *externalLeading = wxMax(0, fe.height - (fe.ascent + fe.descent));
     }
 }
-
+#include <iostream>
```
Comment (iwbnwif) [inline src/generic/graphicc.cpp]: Oops, this shouldn't be here. I will take it out on the next commit.

**Comment by iwbnwif** [comment]
I wrote a little test program based on the Ellipsize test set that draws the resulting text to a bitmap and then blits the bitmap onto a wxScrolledWindow in a paint event. The intention is to test my changes to EllipsizeCalculator.

The red rectangle is the bounding extents of the ellipsized text, whilst the blue rectangle is the allowed width. f:, m:2, w:0 are the indexes into the flags, modes and widths arrays (see the test case for details).

The following images show an extract of the tests with various GDK_DPI_SCALE settings (1.0, 1.25 and 2.0 respectively). Note that the first output in each image looks like it is big enough to contain the text without ellipsizing it, but this is misleading because the text contains a trailing tab and so the 'proper' width of the non-ellipsized text is shown by the red boxes in the second two outputs. I.e. the first output is correctly ellipsized!

GDK_DPI_SCALE 1-0
GDK_DPI_SCALE 1-25
GDK_DPI_SCALE 2-0

**Comment by vadz** [comment]
Great, thanks for fixing this! I think that making a copy of the string isn't a big deal, i.e. this shouldn't be used with megabyte-long strings anyhow, so I'm perfectly fine with this solution.

Do you plan to do anything else here or should this be merged (after some cleanup, e.g. I'd like to at least remove the commented out code0? Thanks!

**Comment by iwbnwif** [comment]
> Do you plan to do anything else here or should this be merged (after some cleanup, e.g. I'd like to at least remove the commented out code? Thanks!

No, I will just tidy up, test a bit more and then post back when I think it is ready for merging.

**Comment by iwbnwif** [comment]
The following two images show the improvement that has been made by the change to EllipsizeCalculator.

In the first image (f1_m2_w0_old_code), the tabs are not ellipsized and therefore the text oveflows the allowed width. The second image (f1_m2_w0_new_code) shows the tabs being correctly ellipsized.

f1_m2_w0_old_code

f1_m2_w0_new_code

**Comment by vadz** [comment]
I'm sorry, but I don't think I agree with the fact that it's an improvement, I don't know if the old behaviour was intentional or not, but I think it makes perfect sense to not count trailing spaces/tabs for the purposes of ellipsization at all. They're invisible anyhow, so why should we truncate the visible string (i.e. remove "g" from it in this example) just to show the ellipsis? But maybe even worse, we can be in a situation when we do _not_ truncate the string and show the ellipsis just because there are some spaces later -- so when the user resizes the column showing this string, ellipsis disappears but nothing is shown instead of it, which certainly looks like a bug.

If the old code already suffered from these problems and they just didn't appear in it for whatever reason, we could still apply this PR and fix this problem later. But if it used to work as expected before and doesn't now, it would be a regression and I'd rather not apply these changes then.

In any case, it hopefully shouldn't be too difficult to just remove the trailing whitespace from each line?

**Comment by iwbnwif** [comment]
> But maybe even worse, we can be in a situation when we do not truncate the string and show the ellipsis just because there are some spaces later -- so when the user resizes the column showing this string, ellipsis disappears but nothing is shown instead of it, which certainly looks like a bug.

As you maybe suspected, the old code would have done this if either there were either trailing spaces or if wxELLIPSIZE_FLAGS_EXPAND_TABS was set. And the old code _may_ also have overflowed the maximum width and failed the test case under those circumstances.

A copy of the line is taken in wxControlBase::Ellipsize, so we can simply wxString::Trim it before passing to the line ellipsizer (DoEllipsizeSingleLine). I wouldn't want to trim it in the ellipsizer itself because it relies (and has always done so) on a correlation between the string length and the character offset array, m_charOffsetsPx.

Implicit in this new change is that it is now policy to trim all white space from the end of each line of ellipsized text before displaying it (the original text will remain unchanged of course). FWIW I think this is a sensible idea and don't see any negative sides, but maybe I am missing something!

**Comment by iwbnwif** [comment]
f1_m2_w0_new_code_with_trim

**Comment by vadz** [comment]
Just FYI: I've created #1649 with the ellipsization fixes and will merge the rest (as a single commit) once the other PR is merged (which I'll do as soon as its CI builds pass).

Thanks for all your fixes!

**Comment by vadz** [comment]
I've extracted the rest of the changes, as well as my own simplification on top of them, into #1650, so I think this one can be closed now.

Please let me know if you see anything wrong with the other PR(s), and if you could please test #1650 with fractional font scaling, it would be great. TIA!

**Comment by iwbnwif** [comment]
Sorry, I have been travelling the past two days and didn't notice that this has now been merged. 

Thank you for the extra work on bringing this to a conclusion. I see there has been some feedback in the reference PR, so I will carry on the conversation there.

**Comment by vadz** [comment]
Just to be clear: this hasn't been merged yet, please leave your comments on #1649 and #1650 if you have any. I do plan to merge them soon if you don't have any objections/remarks. Thanks!

**Pull request closed without merging**